### DISCUSSION TITLE
useAsyncData vs useFetch vs $fetch

### DISCUSSION THREAD

**Discussion by memic84** [nuxt/nuxt#16294] [Questions] [answered]
I just seem to be going back to the documentation on data fetching, and not grasping **when to use** which data fetching method.

It might be related to the fact that i'm still stuck in the Nuxt 2 mindset (were it seemed a bit simpler when to use asyncData and when fetch).

I have read the docs and searched in discussions / issues ... but still miss the answer for **when to use / which use-case**?
I would be open to contribute to the docs, if this could be cleared up.

Some questions and remarks:

1. **useAsyncData**

```js
<script setup>
const { data } = await useAsyncData('count', () => $fetch('/api/count'))
</script>
``` 

_Should be used in more complex situations._ I guess this is when you need to make more calls and resolve them at the same time? 

**Any other examples for this?** 

In Nuxt 2 the asyncData was used just for pages, and blocked the navigation till the promise resolved and fetch in Nuxt 2 could also be used for components.

2. **useFetch**
```js
<script setup>
const { data: count } = await useFetch('/api/count')
</script>
```

Usage:
In most cases this seems the go-to method, when you're facing more simple use-cases. If you're using an external api, creating a useCustomFetch seems the way to go, to define the default values. useFetch wraps useAsyncData and is almost the same. Why have two options for this, which could cause confusion (especially with Nuxt 2 users). Just curious about the decision, because it seems to me, the asyncData could have been the default, and if anybody needed useFetch, you just could have created your own composable.

**Question:**
- **Should useFetch be used for only client side calls, where navigation doesn't need to be blocked?**
- What about the key in options, which is used for caching purposes. Is this mandatory.
- If we need a key, does this mean that the data is globally accessible for all pages and not destroyed when the user leaves the page?
- What if there are multiple components or composables that use useFetch of useAsyncData during SSR and the navigation is blocked, how is this resolved. Does Nuxt wait for all of them to resolve and continue the navigation?

3. **$fetch**
This is the ofetch from unjs which is used, and is also used in useFetch and useAsyncData.
$fetch can be used on the **server** just as on the **client**, but wil never block navigation.

Usage:
- When you need to make client side calls. (eg. fetching filters, or sending login data after input)? (none of the calls are SSR)
- Could be used in middleware, plugins, modules etc?

**Comment by wobsoriano** [comment, marked as answer]
```ts
useFetch('/api/yes', {
   // a unique key to ensure that data fetching can be properly de-duplicated across requests,
   // if not provided, it will be generated based on the static code location where useAsyncData is used.
   key: 'johncena',
   // whether to fetch the data on the server
   server: false
   // whether to resolve the async function after loading the route,
   // instead of blocking client-side navigation (defaults to false)
   lazy: true
})
```

If you want to do API calls in the server, use `$fetch`.

Everything is here 

https://nuxt.com/docs/api/composables/use-async-data
https://nuxt.com/docs/api/composables/use-fetch

**Reply by memic84** [reply]
Thanks! Makes sense, used $fetch in server/api and useFetch in pages/components and it works!

**Reply by kleenkanteen** [reply]
Don't know why this is the accepted answer. You can keep things simple and use useFetch on both the server and client. 99% of the time this will suffice and give no issues. 

The only time you should use useFetch instead of $fetch is when SSR is enabled. This ensures the api called with $fetch is called only once on the server. Otherwise if you use $fetch, the api is called twice, once on the server and another on the client.

I could be wrong but this is what I got. 

Source: https://nuxt.com/docs/api/utils/dollarfetch

**Reply by ronaldrupp** [reply]
@kleenkanteen It's the other way around:

> We recommend to use [useFetch](https://nuxt.com/docs/api/composables/use-fetch) or [useAsyncData](https://nuxt.com/docs/api/composables/use-async-data) + $fetch to prevent double data fetching when fetching the component data.

```
// During SSR data is fetched twice, once on the server and once on the client.
const dataTwice = await $fetch('/api/item')

// During SSR data is fetched only on the server side and transferred to the client.
const { data } = await useAsyncData('item', () => $fetch('/api/item'))

// You can also useFetch as shortcut of useAsyncData + $fetch
const { data } = await useFetch('/api/item')
```

**Reply by kleenkanteen** [reply]
@ronaldrupp thanks, edited my answer

**Comment by aeropagz** [comment]
Hey ,I could need some help too.
What about fetching in composables?
They can be executed on server and cient side, right? So it is not possible to use `useFetch`?  
Also it is not really clear for me what is executed on client or server side. Is there any good documentation? I didn't find any helpful on their page.  

I am very new to nuxt and its quite confusing for me

**Reply by DenFin** [reply]
`useFetch` should only be used in script setup. Anywhere else you should use `$fetch`: https://github.com/nuxt/nuxt/issues/15528#issuecomment-1397383088

**Reply by Jahchap** [reply]
Just to add in, `$fetch` is just a wrapper around `ofetch` which does the request (you can think of `$fetch` as similar to `axios` and you can use `axios` if you want, I have used it myself).

`useAsyncData` is a composable that takes a key and a function that returns a promise, in this case `$fetch` (please note that you can use `axios` as I stated above).

`useFetch`(also a composable) is a shorthand for `useAsyncData` + `$fetch`. See this [link](https://nuxt.com/docs/api/utils/dollarfetch#fetch).

You can use both `useAsyncFetch` and `useFetch` in composables inside the `setup` method or the `script setup` syntax.

Again see @wobsoriano's answer above:

> If you want to do API calls in the server, use `$fetch`.
> 
> Everything is here
> 
> https://nuxt.com/docs/api/composables/use-async-data
> https://nuxt.com/docs/api/composables/use-fetch

**Comment by JamesAlphonse** [comment]
> I just seem to be going back to the documentation on data fetching, and not grasping **when to use** which data fetching method.
> 
> It might be related to the fact that i'm still stuck in the Nuxt 2 mindset (were it seemed a bit simpler when to use asyncData and when fetch).
> 
> I have read the docs and searched in discussions / issues ... but still miss the answer for **when to use / which use-case**? I would be open to contribute to the docs, if this could be cleared up.
> 
> Some questions and remarks:
> 
> 1. **useAsyncData**
> 
> ```js
> <script setup>
> const { data } = await useAsyncData('count', () => $fetch('/api/count'))
> </script>
> ```
> 
> _Should be used in more complex situations._ I guess this is when you need to make more calls and resolve them at the same time?
> 
> **Any other examples for this?**
> 
> In Nuxt 2 the asyncData was used just for pages, and blocked the navigation till the promise resolved and fetch in Nuxt 2 could also be used for components.
> 
> 2. **useFetch**
> 
> ```js
> <script setup>
> const { data: count } = await useFetch('/api/count')
> </script>
> ```
> 
> Usage: In most cases this seems the go-to method, when you're facing more simple use-cases. If you're using an external api, creating a useCustomFetch seems the way to go, to define the default values. useFetch wraps useAsyncData and is almost the same. Why have two options for this, which could cause confusion (especially with Nuxt 2 users). Just curious about the decision, because it seems to me, the asyncData could have been the default, and if anybody needed useFetch, you just could have created your own composable.
> 
> **Question:**
> 
> * **Should useFetch be used for only client side calls, where navigation doesn't need to be blocked?**
> * What about the key in options, which is used for caching purposes. Is this mandatory.
> * If we need a key, does this mean that the data is globally accessible for all pages and not destroyed when the user leaves the page?
> * What if there are multiple components or composables that use useFetch of useAsyncData during SSR and the navigation is blocked, how is this resolved. Does Nuxt wait for all of them to resolve and continue the navigation?
> 
> 3. **$fetch**
>    This is the ofetch from unjs which is used, and is also used in useFetch and useAsyncData.
>    $fetch can be used on the **server** just as on the **client**, but wil never block navigation.
> 
> Usage:
> 
> * When you need to make client side calls. (eg. fetching filters, or sending login data after input)? (none of the calls are SSR)
> * Could be used in middleware, plugins, modules etc?

Of course you do not, I don't either, and the reason is simple: Because for some damn reason the people that make documentation and implement this stuff seem to want to make things more complicated. Just use the normal fetch method.

**Reply by ascorbic-acid** [reply]
fetching data also with ssr on/off is confusing topic , it take time to get it

**Reply by DenFin** [reply]
@ascorbic-acid your statement is very general, so it is hard to give you any specific advice. you could check out some of @manniL's [youtube videos](https://www.youtube.com/@TheAlexLichter/videos). he does a good job explaining Nuxt-specific APIs

**Comment by mihaon** [comment]
> if not provided, it will be generated based on the static code location where useAsyncData is used.

It's wrong in 2026. Currently useFetch key is generated based on url + query params + body + etc. This is a very strange decision that leads to many errors.
